nexus/client.go

```diff
@@ -0,0 +1,629 @@
+func NewExecuteOperationOptions(operation string, v any) (options ExecuteOperationOptions, err error) {
+	if !isValidOperationName.MatchString(operation) {
+		err = errInvalidOperationName
+		return
+	}
+	var b []byte
+	b, err = json.Marshal(v)
+	if err != nil {
+		return
+	}
+	options.Operation = operation
+	options.StartHeader = http.Header{headerContentType: []string{contentTypeJSON}}
+	options.Body = bytes.NewReader(b)
+	return
+}
+
+func (r *ExecuteOperationOptions) intoStartOptions() StartOperationOptions {
+	return StartOperationOptions{
+		Operation:   r.Operation,
+		CallbackURL: r.CallbackURL,
+		RequestID:   r.RequestID,
+		Header:      r.StartHeader,
+		Body:        r.Body,
+	}
+}
+
+func (r *ExecuteOperationOptions) intoGetResultOptions() (options GetResultOptions) {
+	options.Header = r.GetResultHeader
+	if r.Wait <= 0 {
+		options.Wait = infiniteDuration
+	} else {
+		options.Wait = r.Wait
+	}
+	return
+}
+
+// ExecuteOperation is a helper for starting an operation and waiting for its completion.
+//
+// For asynchronous operations, the client will long poll for their result, issuing one or more requests until the
+// wait period provided via [ExecuteOperationOptions] exceeds, in which case an [ErrOperationStillRunning] error is
+// returned.
+//
+// The wait time is capped to the deadline of the provided context.
+//
+// ⚠️ If this method completes successfully, the returned response's body must be read in its entirety and closed to
+// free up the underlying connection.
+func (c *Client) ExecuteOperation(ctx context.Context, request ExecuteOperationOptions) (*http.Response, error) {
+	result, err := c.StartOperation(ctx, request.intoStartOptions())
+	if err != nil {
+		return nil, err
+	}
+	if result.Successful != nil {
+		return result.Successful, nil
+	}
+	handle := result.Pending
+	return handle.GetResult(ctx, request.intoGetResultOptions())
+}
+
+// NewHandle gets a handle to an asynchronous operation by name and ID.
+// Does not incur a trip to the server.
+// Fails if provided any empty operation or ID, or the client has no associated ServiceBaseURL.
+func (c *Client) NewHandle(operation string, operationID string) (*OperationHandle, error) {
+	var es []error
+	if c.serviceBaseURL == nil {
+		es = append(es, errEmptyServiceBaseURL)
+	}
+	if !isValidOperationName.MatchString(operation) {
+		es = append(es, errInvalidOperationName)
+	}
+	if !isValidOperationID.MatchString(operationID) {
+		es = append(es, errInvalidOperationID)
+	}
+	if len(es) > 0 {
+		return nil, errors.Join(es...)
+	}
+	return &OperationHandle{
+		client:    c,
+		Operation: operation,
+		ID:        operationID,
+	}, nil
+}
+
+func (c *Client) sendGetOperationResultRequest(ctx context.Context, httpReq *http.Request, wait time.Duration) (*http.Response, error) {
+	if wait > 0 {
+		timeout := min(wait, c.Options.GetResultMaxTimeout)
+		if deadline, set := ctx.Deadline(); set {
+			timeout = min(timeout, time.Until(deadline))
```

Comment:
> Adding a bit to context deadline here ensures that context deadline exceeded is always the error a client sees on context deadline exceeded.

Not if the timeout is roughly equal to to until deadline + some buffer.
Let's say the deadline is 40 seconds and timeout is 60, you'll still race with your suggestion.

At the end of the day if you have a deadline, it's better to use that and not set a wait duration to avoid the ambiguity. Otherwise, set the wait duration.

Reply:
I'm starting to think that we need to just rely on context deadline and expose `Wait` as a bool

Reply:
No, you're going backwards. You are conflating two issues: protocol timeout (i.e. client-side HTTP _call_ timeout via context) and application timeout (i.e. get result returns is-running server side via wait). When you confuse these two things all you get is trouble. The only time one should affect the other is if you want to help the server by only giving a wait a bit past the protocol timeout since the client will go away anyways.

Also note that you have server side timeout which is a third thing that should also be separate. That value is just when the server may return regardless of wait and this should retry.

Basically, these scenarios and their call counts should be deterministic:

* Protocol timeout: 5s, wait: 5s, server timeout: 8s - what is the error returned on timeout? Should be protocol timeout _always_ and only one call.
* Protocol timeout: 5s, wait: 10s, server timeout: 8s - what is the error returned on timeout? Should be protocol timeout _always_ and only one call.
* Protocol timeout: 10s, wait: 5s, server timeout: 8s - what is the error returned on timeout? Should be application timeout _always_ and only one call.
* Protocol timeout: 20s, wait: 10s, server timeout: 8s - what is the error returned on timeout? Should be application timeout _always_ and two calls (after 8s, server returns with server-timeout error which is _not_ the same as still-running, and then you call again with a wait of 2s).

For the sake of unit tests and clarity, we should not confuse these. There's no good reason. Contexts are often set outside the caller anyways. The only way it is going to work clearly and sanely is if you don't let protocol timeouts affect wait timeouts and you differentiate server timeout from application timeout.

Reply:
We do have a server max timeout but it prefers to the client provided timeout.
The server will need to give some headroom to allow requests to complete, e.g. if a client sets wait to 60 seconds, the server should allow up to 60 - headroom seconds for the request.
The server should always respond with not-ready-yet (precondition failed) when the allowed time is over, but it will wait for the handler to complete processing the request.

> Protocol timeout: 5s, wait: 8s, server timeout: 10s - Should be protocol timeout to user always and only one call

I disagree, the application is willing to wait another 3 seconds, why not issue a shorter call with a 3 second timeout?
If the remaining wait time is shorter than what we consider reasonable request timeout, we could avoid sending out a request, but that value should be configurable on the client.

As for your previous concern for context deadline that's set to the same value as the wait deadline (now + timeout), if we want consistent deadline exceeded errors, we should use the context deadline and avoid this situation altogether, otherwise there'll always be racy situations.

Reply:
> > Protocol timeout: 5s, wait: 8s, server timeout: 10s - Should be protocol timeout to user always and only one call

> I disagree, the application is willing to wait another 3 seconds, why not issue a shorter call with a 3 second timeout?

The application _is not_ willing to wait another three seconds because their context timeout is 5s. That's it, context done, call done. When I say "protocol timeout" I am talking about the context deadline. But at least they'll always get a deadline exceeded error in this _very common_ scenario. Much better than this _very common_ scenario having unclear results.

> if we want consistent deadline exceeded errors, we should use the context deadline and avoid this situation altogether, otherwise there'll always be racy situations.

Only when users set the same values will there always be racy situations. When they don't, it should be clear. _We_ shouldn't cause racy situations in the normal case of protocol timeout (context deadline) being shorter than application timeout (wait).

I think my 7 scenarios above are very clear behavior.